File: go/dsa/quicksort.go
package dsa

func quicksort(arr []int) {
	quicksortHelper(arr, 0, len(arr)-1)
}

func quicksortHelper(arr []int, lo, hi int) {
	if lo < hi {
		p := partition(arr, lo, hi)
		quicksortHelper(arr, lo, p-1)
		quicksortHelper(arr, p+1, hi)
	}
}

func partition(arr []int, lo, hi int) int {
	pivot := arr[hi]
	partition := lo - 1
	for j := lo; j < hi; j++ {
		if arr[j] < pivot {
			partition += 1
			swap(arr, j, partition)
		}
	}
	partition += 1
	swap(arr, partition, hi)
	return partition
}

func swap(arr []int, j, i int) {
	temp := arr[j]
	arr[j] = arr[i]
	arr[i] = temp
}

File: typescript/algorithms/quicksort.ts
export function quicksort(nums: number[]) {
  if(nums.length <= 1) return nums;
  const p = partition(nums);
  return quicksort(nums.slice(0, p)).concat(quicksort(nums.slice(p)))
}

function partition(nums: number[]): number {
  const pivot = nums[nums.length-1];
  let p = nums.length;
  let i = 0;
  while(i < p) {
    if(nums[i] > pivot) {
      --p;
      const temp = nums[p];
      nums[p] = nums[i];
      nums[i] = temp;
    }
    i++;
  }
  return p;
}

console.log(quicksort([4,2,6,3,8,5]))